crates/ruff_linter/src/rules/ruff/rules/abstract_method_in_normal_class.rs

```diff
@@ -0,0 +1,147 @@
+use ruff_python_ast::{Stmt, StmtClassDef, StmtFunctionDef};
+use ruff_python_semantic::analyze::visibility::{abstract_decorator, AbstractDecoratorKind};
+
+use crate::checkers::ast::Checker;
+
+/// ## What it does
+/// Checks for methods decorated with at least one of `abc`'s
+/// `abstractmethod`, `abstractclassmethod`, `abstractstaticmethod` and `abstractproperty`
+/// but defined within a normal class's body.
+///
+/// ## Why is this bad?
+/// The abstract method decorators prevent users from instantiating abstract classes,
+/// or inheriting from abstract classes without implementing all abstract methods
+/// by throwing an exception. Such abstract method decorators are only effective
+/// in an abstract class.
+///
+/// For a mixin class, it is not enough that `abc.ABC` is included in the eventual MRO.
+/// The mixin class must also inherit directly from `ABC` for the decorators to take effect.
+///
+/// ```python
+/// from abc import ABC, abstractmethod
+///
+///
+/// class Base(ABC):
+///     @abstractmethod
+///     def hello(self) -> None: ...
+///
+///     def __repr__(self) -> str:
+///         return f"message={self.msg!r}"
+///
+///
+/// class Mixin:  # should be: `Mixin(ABC)`:
+///     @abstractmethod
+///     def world(self) -> None:
+///         self.msg += " goodbye"
+///
+///
+/// class FooBar(Mixin, Base):
+///     def __init__(self):
+///         self.msg = ""
+///
+///     def hello(self) -> None:
+///         self.msg += "hello"
+///
+///     # without `Mixin(ABC)`, omitting this does not raise an exception
+///     # def world(self) -> None:
+///     #     self.msg += " world"
+///
+///
+/// # `ABC` is part of the MRO
+/// print(FooBar.mro())  # [FooBar, Mixin, Base, ABC, object]
+///
+/// fb = FooBar()
+/// fb.hello()
+/// fb.world()
+/// print(str(fb))  # message='hello goodbye'
+/// ```
+///
+/// ## Example
+///
+/// ```python
+/// class C:
+///     @abstractmethod
+///     def m(self) -> None:
+///         pass
+/// ```
+///
+/// Use instead:
+///
+/// ```python
+/// from abc import ABC
+///
+///
+/// class C(ABC):
+///     @abstractmethod
+///     def m(self) -> None:
+///         pass
+/// ```
+#[derive(ViolationMetadata)]
+pub(crate) struct AbstractMethodInNormalClass {
+    decorator_kind: AbstractDecoratorKind,
+    class_name: String,
+}
+
+impl Violation for AbstractMethodInNormalClass {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        "Abstract method defined in non-abstract class".to_string()
+    }
+
+    fn fix_title(&self) -> Option<String> {
+        let (decorator, class) = (&self.decorator_kind, &self.class_name);
+
+        Some(format!(
+            "Remove `@{decorator}` or make `{class}` inherit from `abc.ABC`"
+        ))
+    }
+}
+
+/// RUF044
+pub(crate) fn abstract_method_in_normal_class(checker: &mut Checker, class: &StmtClassDef) {
+    if checker.source_type.is_stub() {
+        return;
+    }
+
+    if !class.bases().is_empty() {
+        return;
+    }
```

Comment:
Should we also check the base classes that are in the same file? I'm fine with moving forward but this seems like something that's doable. I'm specifically referring to point (1) that's mentioned https://github.com/astral-sh/ruff/issues/12861#issuecomment-2288460681:

> We need to recognise as valid any class that has metaclass=X, or any class that inherits from a class that has metaclass=X, where X is either abc.ABCMeta or a subclass of abc.ABCMeta.

Reply:
I'm not against doing this as a future extension for as long as it doesn't result in the rule having too many false negatives. I'm already somewhat concerned that the rule has too many false negatives because of Ruff's lack for multifile analysis.

Reply:
What I'm saying is that it depends. I'm worried that the rule has too many false positives to be useful because it only catches a small subset of violations. Supporting classes with base classes in the same file could help to reduce the number of false positives. 

I'm not sure how much effort it is but it would be nice to see how many additional violation the rule catches in the ecosystem check if it had support for base classes in the same file. Do you think it would be possible to *hack* something together just to get that data? It doesn't have to be a 100% correct.

Looking at it from a different angle, an alternative is to keep it as is and document it as a known limitation that the rule only applies to classes without base classes. The benefit of this is that it's a very simple limitation that's easy understood